src/Features/LanguageServer/Protocol/Handler/SourceGenerators/GetTextHandler.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [ExportCSharpVisualBasicStatelessLspService(typeof(GetTextHandler)), Shared]
+    [Method("sourceGenerator/getText")]
+    internal class GetTextHandler : ILspServiceDocumentRequestHandler<SourceGeneratorGetTextParams, SourceGeneratedDocumentText>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public GetTextHandler()
+        {
+        }
+
+        public bool MutatesSolutionState => false;
+        public bool RequiresLSPSolution => true;
+
+        public LSP.TextDocumentIdentifier GetTextDocumentIdentifier(SourceGeneratorGetTextParams request) => request.TextDocument;
```

Comment:
1. interesting, so we support getting in that uri for the SG doc in the handler itself, and having that map to the Doc that is in the context object?  nice.
2. i'm def not quite understanding the threading/generation model here.  Specifically, the client could certainly ask us about a doc that we're no longer gnerating right?  If that happens, what's the 'fail' model?  Do we throw deep in the server message processing?  do we bail gracefully and not execute the request at all?  do we synthesize a dummy empty doc?  can you clarify?

Reply:
> interesting, so we support getting in that uri for the SG doc in the handler itself, and having that map to the Doc that is in the context object? nice.

Yep, the generated document URI is passed for any document request and we'll return it like anything else. So all other features should hopefully light up with no surprises here.

> i'm def not quite understanding the threading/generation model here. Specifically, the client could certainly ask us about a doc that we're no longer gnerating right? If that happens, what's the 'fail' model? Do we throw deep in the server message processing? do we bail gracefully and not execute the request at all? do we synthesize a dummy empty doc? can you clarify?

This part isn't implemented in this PR yet (it's still draft! :smile:) But the source generated files we get still produce didOpen/didChange/didClose events like any other opened document. So just like we update the LSP solution to match all the text changes that came in via didChange for regular documents, we can do the same here. We already have support in the core workspace for "treat this source generated file as having this other content" and "force this document to still exist, regardless". So in that case, if the file is still open but no longer being generated, we can just give a solution snapshot to the handler that still contains that document with the client's text. The other option is we just count it as a miscellaneous file at that point. So no matter what, the handler can run, but the document is not quite real in one way or another.